NuttX using Wokwi (#182)
espressif/developer-portal: content/blog/nuttx-using-wokwi/index.md

---

> title: "Using Wokwi with NuttX"

Reviewer: Should we move to the `content/2024/11/` folder?

> title: "Using Wokwi with NuttX"

Author: I didn't get the exact directory system. Do you mean something like this?:

.
└── developer-portal/
    └── content/
        └── 2024/
            └── 11/
                └── using-wokwi-with-nuttx/
                    ├── img
                    ├── index.md
                    └── ...

> title: "Using Wokwi with NuttX"

Reviewer: Yes! The folder `2024` is already there, please rebase to get the latest and add the `11` folder.

## Alternatives and Differences

If we compare Wokwi with QEMU, which is not the only alternative but one of the most known one. **QEMU** stands out with faster operation with special optimizations while **Wokwi** offers more platform support, ease of installation and simulation features with external parts.

## Simulating NuttX on Wokwi

### Creating a NuttX System Image

  - You can create the development environment for Xtensa-based (ESP32, ESP32-S2, ESP32-S3) microcontrollers using [this link](https://nuttx.apache.org/docs/latest/platforms/xtensa/index.html).
  - You can create the development environment for RISC-V based (ESP32-C3, ESP32-C6, ESP32-H2) microcontrollers using [this link](https://nuttx.apache.org/docs/latest/platforms/risc-v/index.html).

Wokwi does not support compiling NuttX projects internally like ESP-IDF, Zephyr. Therefore, it is necessary to compile locally and continue the simulation part via Wokwi. In other words, we will simulate the NuttX output compiled in our own development environment via Wokwi.

> In this article, we will use the nsh application for the esp32 microcontroller as an example. Enter the following commands to clean if there is a previously selected configuration and to select configuration needed for nsh application for esp32.

Author: It is the most basic application on NuttX. I don't think we need it.

```bash
make distclean
./tool/configure.sh esp32-devkitc:nsh
```

With these commands, we have selected the necessary card and configuration. To perform debugging, we need to activate the `Generate Debug Symbols` feature in the `Debug Options` section of the `Build System` menu. We can use the following command for the relevant setting:

```bash
make menuconfig
```

The result will look something like this:

> ![make menuconfig](img/make_menuconfig.webp)

Reviewer 2: Suggest that you make the window smaller and then take the screenshot, because the valuable information is the text in the middle which is very small.

After configuration and setup, we can use the following command to compile:

```bash
make
```

We should see the following output:

```text
LD: nuttx
CP: nuttx.hex
CP: nuttx.bin
```

The `nuttx` file will be the file we need in the simulation after the compilation process is successfully completed.

### Simulating NuttX Application in Wokwi

> To use the Wokwi with Visual Studio Code, you must first download the [`Wokwi for VS Code` ](https://marketplace.visualstudio.com/items?itemName=wokwi.wokwi-vscode) plug-in and enter the license key. You can get more detailed information about it by using [this link](https://docs.wokwi.com/vscode/getting-started).

Reviewer 2: ```suggestion
To use Wokwi with Visual Studio Code, you must first download the [`Wokwi for VS Code` ](https://marketplace.visualstudio.com/items?itemName=wokwi.wokwi-vscode) plug-in and enter the license key. You can get more detailed information about it [here](https://docs.wokwi.com/vscode/getting-started).
```

> Wokwi basically works on two files. These are `wokwi.toml` and `diagram.json` files. If we give more information about these files:

Reviewer 2: ```suggestion
Wokwi basically works on two files - `wokwi.toml` and `diagram.json`:
```

---

> title: "Nuttx Using Wokwi"

Reviewer: Change the title to: Using Wokwi with NuttX

date: 2024-10-30T00:08:20+01:00
tags: ["NuttX", "Apache", "ESP32", "POSIX", "Linux", "Wokwi", "Simulation"]
showAuthor: false
featureAsset: "img/featured_nuttx_using_wokwi.webp"
authors:
    - "eren-terzioglu"
---

> Note: Reviewing this article together with <cite>[NuttX on an emulated ESP32 using QEMU][1]</cite> article will help the user to distinguish between the two solutions and to determine the most suitable solution for the needs. Additionally, it is recommended to have knowledge of project compilation through the NuttX official documentation and <cite>[Getting Started with NuttX and ESP32][7]</cite> article. NuttX project build documentation is available [at this link](https://nuttx.apache.org/docs/latest/).

Reviewer: The note shouldn't be in the very beginning. Please move it after the introduction.

> ## Using Wokwi with NuttX

Reviewer: Rename to Introduction or leave without heading.

In this article, we will examine how the NuttX real-time operating system (RTOS) can be used on the Wokwi simulator, what adjustments need to be made on the NuttX side for this and what advantages the Wokwi simulator offers.

> ## What is Wokwi

Reviewer: What is Wokwi?

<cite>[Wokwi is an online Electronics simulator. You can use it to simulate Arduino, ESP32, STM32, and many other popular boards, parts and sensors.][2]</cite>

> ## What are the Advantages of Wokwi

Reviewer: If you are going to answer, add the question mark.

  - It provides the opportunity to start the project regardless of the delivery time of the parts required for the project.

  - There is no way to destroy virtual hardware. For that reason cash loss due to connection error is prevented and it provides the opportunity to undo the transaction after a step goes wrong.

  - Provides convenience in getting help and feedback through the link to the project.

  - Helps distinguish hardware and software based problems. This way, you can focus on the code side.

  - It allows the use of any number of hardware regardless of the project budget.

<cite>[Source][3]</cite>

Note: Of course, there will be differences between the simulation and the field. For this reason, the project running in simulation should not be taken to the field without testing.

## Supported devices

> 

Reviewer: I suggest adding some text, like:

Here is the list of supported targets, separated by the core architecture.

>   ###### Xtensa

Reviewer: Change to a table with 3 columns: SoC, Core, and Support Stage.

    - ESP32
    - ESP32-S2
    - ESP32-S3
  ###### RISC-V
    - ESP32-C3
    - ESP32-C6
    - ESP32-H2*
    - ESP32-P4**

<cite>`*`: In beta stage</cite>

<cite> `**`: In alpha stage</cite>

<cite>[Source][4]</cite>. Apart from these, you can use [this link](https://docs.wokwi.com/getting-started/supported-hardware) for supported sensors and more.

> ## Alternatives to Wokwi and What Are the Differences Between These Alternatives

Reviewer: The section title is too long.

Simulation: Differences and Alternatives

If we compare Wokwi with QEMU, which is not the only alternative but one of the most known one. **QEMU** stands out with faster operation with special optimizations while **Wokwi** offers more platform support, ease of installation and simulation features with external parts.

> ## Simulating NuttX Real-Time Operating System on Wokwi

Reviewer: Simulating NuttX on Wokwi

### Creating a NuttX System Image

> Note: This section assumes that the NuttX development environment is installed. You can refer to [NuttX official documentation](https://nuttx.apache.org/docs/latest/) and <cite>[Getting Started with NuttX and ESP32][7]</cite> article for installation of the development environment. If you have already installed the development environment, do not forget to check if it is up to date. For the development environment for Espressif devices:

Reviewer: Remove Note:.

  - You can create the development environment for Xtensa-based (ESP32, ESP32-S2, ESP32-S3) microcontrollers using [this link](https://nuttx.apache.org/docs/latest/platforms/xtensa/index.html).
  - You can create the development environment for RISC-V based (ESP32-C3, ESP32-C6, ESP32-H2) microcontrollers using [this link](https://nuttx.apache.org/docs/latest/platforms/risc-v/index.html).

Wokwi does not support compiling NuttX projects internally like ESP-IDF, Zephyr. Therefore, it is necessary to compile locally and continue the simulation part via Wokwi. In other words, we will simulate the NuttX output compiled in our own development environment via Wokwi.

In this article, we will use the nsh application for the esp32 microcontroller as an example. Enter the following commands to clean if there is a previously selected configuration and to select configuration needed for nsh application for esp32.

> ```

Reviewer: Do not leave without the type when adding code snippets.
You can use `bash`, `text`, `c`, etc.

make distclean
./tool/configure.sh esp32-devkitc:nsh
```

With these commands, we have selected the necessary card and configuration. To perform debugging, we need to activate the `Generate Debug Symbols` feature in the `Debug Options` section of the `Build System` menu. We can use the following command for the relevant setting:

> ```

Reviewer: Same for all.

make menuconfig
```

The result will look something like this:

> ![make menuconfig](img/make_menuconfig.png)

Reviewer: Change it to the webp file.

Wokwi basically works on two files. These are `wokwi.toml` and `diagram.json` files. If we give more information about these files:

- **wokwi.toml:** <cite>[A configuration file that tells Wokwi how to run your project.][5]</cite> Information such as 'elf' and 'firmware' file paths of the application are saved in this file.

- **diagram.json:** <cite>[A diagram file that describes the circuit.][5]</cite> Information such as the microcontroller to be added to the circuit, peripheral units such as the sensor to be connected to the microcontroller, their locations, color and interconnection of the parts in the circuit are recorded in this file. You can use [this link](https://docs.wokwi.com/diagram-format) for detailed information.

For more detailed information about Wokwi configuration files, you can use [this link](https://docs.wokwi.com/vscode/project-config).

If we go back to the Simulation section after these definitions, we need to add the `nuttx` file path to the `wokwi.toml` file and also we need the circuit diagram that will use in simulation to the `diagram.json` file. The file contents we need in this example project will be as follows:

> ###### wokwi.toml

Reviewer: Change to `**wokwi.toml**`

```
[wokwi]
version = 1
firmware = 'nuttx'
elf = 'nuttx'
```

> ###### diagram.json

Reviewer: Same here.

```
{
  "version": 1,
  "author": "Anonymous maker",
  "editor": "wokwi",
  "parts": [
    { "type": "board-esp32-devkit-c-v4", "id": "esp", "top": 0, "left": 0, "attrs": {} } }
  ],
  "connections": [
    [ "esp:TX", "$serialMonitor:RX", "", [] ],
    [ "esp:RX", "$serialMonitor:TX", "", [] ]
  ],
  "dependencies": {}
}
```

After creating the necessary files, we can start the simulation by pressing `F1` and the selecting the `Wokwi: Start Simulator` option:

> ![start simulator](img/start_simulator.png)

Reviewer: Change it to the webp file.

After that, Wokwi will be up and running and you can start using NuttX.

> ![NuttX help](img/wokwi_help.png)

Reviewer: Change it to the webp file and the same for all `.png`.

      "type": "cppdbg",
      "request": "launch",
      "program": "${workspaceFolder}/nuttx/nuttx",
      "cwd": "${workspaceFolder}",
      "MIMode": "gdb",
      "miDebuggerPath": "/debugger/path/xtensa-esp32-elf-gdb",
      "miDebuggerServerAddress": "localhost:3333"
    }
  ]
}
```

After that, press `F1` and select `Wokwi: Start Simulator and Wait for Debugger` to start the simulator. Wokwi will then go into standby mode to allow the debugger to connect. In a separate terminal tab you can run GDB with `xtensa-esp32-elf-gdb` or similar commands for Xtensa devices or `riscv-none-elf-gdb` for RISC-V devices. After running the debugger, you can start to debug with Wokwi using the `F5` key in simulator.

For more detailed information about the debugging, you can visit [this link](https://docs.wokwi.com/guides/debugger).

Of course, our hope is that you will not need this process :)

## Conclusion

In this article, we have seen the features of Wokwi and how it can be used with NuttX. Of course, all the features of Wokwi are not limited to these. You can use [this link](https://docs.wokwi.com/?utm_source=wokwi) to get more detailed information and examples about Wokwi.

> ## Useful Links

Reviewer: Change to `Resources`
